File: client/common/clientstream.go
package common

import (
	"net"

	log "github.com/sirupsen/logrus"
)

// ClientStream Interface to interact with the client socket
type ClientStream struct {
	conn  net.Conn
	ID    string
	msgID int
}

// NewClientStream Initializes client socket. In case of
// failure, error is printed in stdout/stderr and exit 1
// is returned
func NewClientStream(ID string, ServerAddress string) *ClientStream {
	conn, err := net.Dial("tcp", ServerAddress)
	if err != nil {
		// Exits the program
		log.Fatalf(
			"action: connect | result: fail | client_id: %v | error: %v",
			ID,
			err,
		)
	}
	log.Infof("action: connect | Connected to server on %v", ServerAddress)
	stream := &ClientStream{
		conn:  conn,
		ID:    ID,
		msgID: 1,
	}
	return stream
}

// Send Sends data in byte format to the server preventing short writes
func (c *ClientStream) Send(data []byte) error {
	for totalSent := 0; totalSent < len(data); {
		sent, err := c.conn.Write(data[totalSent:])
		if err != nil {
			log.Errorf("action: send_message | result: fail | client_id: %v | error: %v", c.ID, err)
			return err
		}
		totalSent += sent
	}
	log.Debugf("action: send_message | result: success | client_id: %v | encoded msg: %v", c.ID, data)
	c.msgID++
	return nil
}

// Recv Receives data from the server and returns it in byte format preventing short reads
func (c *ClientStream) Recv(size int) ([]byte, error) {
	data := make([]byte, size)
	for totalRecvBytes := 0; totalRecvBytes < size; {
		recvBytes, err := c.conn.Read(data[totalRecvBytes:])
		if err != nil {
			log.Errorf("action: receive_message | result: fail | client_id: %v | error: %v", c.ID, err)
			return data, err
		}
		totalRecvBytes += recvBytes
	}
	log.Debugf("action: receive_message | result: success | client_id: %v | encoded msg: %v", c.ID, data)
	return data, nil
}

// Close Closes the connection to the server
func (c *ClientStream) Close() {
	log.Info("action: closing_socket | Closing client socket")
	c.conn.Close()
}

File: server/common/client_stream.py

import logging
import socket

from common.exceptions import SocketConnectionBroken


class ClientStream:
    """
    Interface to send and receive from a socket and prevent short reads/writes
    when communicating with the client
    """

    def __init__(self, socket: socket):
        """
        socket must be a TCP socket
        """
        self.socket = socket
        self.addr = socket.getpeername()

    def send(self, data):
        """ 
        Sends byte data to the client preventing short sends
        Raises SocketConnectionBroken if there is an error
        """
        total_bytes = len(data)
        total_sent = 0
        while total_sent < total_bytes:
            bytes_sent = self.socket.send(data[total_sent:])
            if bytes_sent == 0:
                logging.error(
                    f"action: send_message | result: fail | ip: {self.addr[0]} | No bytes sent, socket connection broken")
                raise SocketConnectionBroken()
            total_sent += bytes_sent
        logging.debug(
            f"action: sent_message | result: success | ip: {self.addr[0]} | msg: {data}")

    def recv(self, total_recv: int):
        """ 
        Receives and returns byte data from the client preventing short reads.
        Raises SocketConnectionBroken if there is an error
        """
        recv_bytes = 0
        data = b""
        while recv_bytes < total_recv:
            recv_data = self.socket.recv(total_recv - recv_bytes)
            if recv_data == b'':
                logging.error(
                    f"action: receive_message | result: fail | ip: {self.addr[0]} | No bytes received, socket connection broken")
                raise SocketConnectionBroken()
            recv_bytes += len(recv_data)
            data += recv_data
        logging.debug(
            f"action: receive_message | result: success | ip: {self.addr[0]} | msg: {data}")
        return data

    def close(self):
        """ Closes the socket """
        logging.info(
            f"action: closing_socket | Closing client socket {self.addr[0]}")
        self.socket.close()
